### PR TITLE
Allow modifying or deleting last message and by Mattermost message ID

### PR DESCRIPTION
**Pull request by hloeung**
Store the Mattermost message ID of the most recent message sent to channel. With that, we can use 'last' to modify or delete. e.g.

|21:15 <hloeung> Test
|21:15 <hloeung> s//Update last test message
|21:15 <hloeung> s//

Also add support for modifying messages using the Mattermost message ID. e.g.

|21:14 <hloeung> Test [@@o5ctj8u3qfgtxf5icisewtdyty]
|21:14 <hloeung> s/o5ctj8u3qfgtxf5icisewtdyty/Update
|21:14 <hloeung> Update (edited) [@@o5ctj8u3qfgtxf5icisewtdyty]
|21:14 <hloeung> s/o5ctj8u3qfgtxf5icisewtdyty/
|21:14 <hloeung> Update (deleted) [@@o5ctj8u3qfgtxf5icisewtdyty]

### CODE REVIEW

**Comment by axinojolais** [comment]
Should there be any documentation update to go along with this change ?

**Comment by hloeung** [comment]
> Should there be any documentation update to go along with this change ?

Updated docs with how to update/modify or delete.

**Comment by hloeung** [comment]
Would I need a mutex with this change?

**Comment by axinojolais** [comment]
Hello, will this be merged any time soon ? Thanks !

**Comment by hloeung** [comment]
@42wim, free to take a look at this one?

Thanks!

**Comment by hloeung** [comment]
That last commit is what @jetpackdanger saw with just PrefixContext/SuffixContext. Thought I'll bundle the fix with this PR, happy to separate it out if you want @42wim.

**Comment by 42wim** [comment]
I'm testing this, but the `s//` or `s//blah` only works one time, then it stops working.
You don't have this issue?

**Comment by hloeung** [comment]
> I'm testing this, but the `s//` or `s//blah` only works one time, then it stops working.
> You don't have this issue?

@42wim, Ah sorry, 9a80c59 broke it. I've fixed it now.

**Comment by hloeung** [comment]
|01:54 <hloeung> Test from web UI [@@f4fc…]
    |01:54 <hloeung> s/f4fck99awfgg98crxpturrzwdr/Modify post from web UI
    |01:54 <hloeung> Modify post from web UI (edited) [@@f4fc…]
    |01:55 <hloeung> Test new one
    |01:55 <hloeung> s//Modify the last 'Test new one'
    |01:55 <hloeung> Test
    |01:55 <hloeung> s//
    |01:55 <hloeung> Test
    |01:55 <hloeung> s//Test update
    |01:55 <hloeung> s//Test update again

Produces:


U58YvLLlvQ

**Comment by hloeung** [comment]
Also with matterircd generated message/thread IDs:

    |02:07 <hloeung> Test from web UI [007]
    |02:07 <hloeung> s/007/Update message from web UI
    |02:07 <hloeung> Update message from web UI (edited) [007]
    |02:07 <hloeung> s//Update again
    |02:07 <hloeung> Update again (edited) [007]
    |02:07 <hloeung> New message
    |02:07 <hloeung> s//

(yes, it really is 2am here heh)

y4Pf0Nxewk

File: mm-go-irckit/server_commands.go [hunk 1]
```diff
@@ -384,6 +384,10 @@ func CmdPrivMsg(s Server, u *User, msg *irc.Message) error {
 		msgID, err2 := u.br.MsgChannel(ch.ID(), msg.Trailing)
 		if err2 != nil {
 			u.MsgSpoofUser(u, u.br.Protocol(), "msg: "+msg.Trailing+" could not be send: "+err2.Error())
+		} else {
+			u.msgLastMutex.Lock()
+			defer u.msgLastMutex.Unlock()
+			u.msgLast[ch.ID()] = msgID
```
Comment (42wim) [inline mm-go-irckit/server_commands.go:390]:
```suggestion
			return err2
		}
		
		u.msgLastMutex.Lock()
		defer u.msgLastMutex.Unlock()
		
		u.msgLast[ch.ID()] = msgID
```

Else isn't needed there, we should return first (not your fault, I should've done that)

File: mm-go-irckit/server_commands.go [hunk 2]
```diff
@@ -449,29 +456,51 @@ func parseModifyMsg(u *User, msg *irc.Message, channelID string) bool {
...
+	switch {
+	// last message from user. Also support '!!' like shell's history
+	// substitution for previous command.
+	case matches[1] == "//" || matches[1] == "/!!/":
+		u.msgLastMutex.RLock()
+		defer u.msgLastMutex.RUnlock()
+		msgLast, ok := u.msgLast[channelID]
+		if ok {
+			msgID = msgLast
+		}
```
Comment (42wim) [inline mm-go-irckit/server_commands.go:470]:
```suggestion
		if msgLast, ok := u.msgLast[channelID];ok {
			msgID = msgLast
		}
```

File: mm-go-irckit/server_commands.go [hunk 3]
```diff
@@ -449,29 +456,51 @@ func parseModifyMsg(u *User, msg *irc.Message, channelID string) bool {
...
-	for k, v := range m {
-		if v == int(id) {
-			msgID = k
+		for k, v := range m {
+			if v == int(id) {
+				msgID = k
+				u.msgLastMutex.Lock()
+				defer u.msgLastMutex.Unlock()
+				u.msgLast[channelID] = msgID
+			}
```
Comment (42wim) [inline mm-go-irckit/server_commands.go:495]:
```suggestion
			if v != int(id) {
				continue
			}
			
			msgID = k
			
			u.msgLastMutex.Lock()
			defer u.msgLastMutex.Unlock()
			
			u.msgLast[channelID] = msgID
```

Original loop was only one statement, now we have more statement this is better.

File: mm-go-irckit/server_commands.go [hunk 4]
```diff
@@ -528,23 +557,39 @@ func threadMsgChannel(u *User, msg *irc.Message, channelID string) bool {
...
 	if err != nil {
 		u.MsgSpoofUser(u, u.br.Protocol(), "msg: "+text+" could not be send: "+err.Error())
+	} else {
+		u.msgLastMutex.Lock()
+		defer u.msgLastMutex.Unlock()
+		u.msgLast[channelID] = msgID
+	}
+
+	if u.v.GetBool(u.br.Protocol()+".prefixcontext") || u.v.GetBool(u.br.Protocol()+".suffixcontext") {
+		u.prefixContext(channelID, msgID, "", "")
```
Comment (42wim) [inline mm-go-irckit/server_commands.go:570]:
```suggestion
		return false
	}

	u.msgLastMutex.Lock()
	defer u.msgLastMutex.Unlock()
	
	u.msgLast[channelID] = msgID

	if u.v.GetBool(u.br.Protocol()+".prefixcontext") || u.v.GetBool(u.br.Protocol()+".suffixcontext") {
		u.prefixContext(channelID, msgID, "", "")
```

File: mm-go-irckit/server_commands.go [hunk 5]
```diff
@@ -528,23 +557,39 @@ func threadMsgChannel(u *User, msg *irc.Message, channelID string) bool {
...
 
-	_, err := u.br.MsgUserThread(toUser, msgID, text)
+	msgID, err := u.br.MsgUserThread(toUser, msgID, text)
 	if err != nil {
 		u.MsgSpoofUser(u, u.br.Protocol(), "msg: "+text+" could not be send: "+err.Error())
+	} else {
+		u.msgLastMutex.Lock()
+		defer u.msgLastMutex.Unlock()
+		u.msgLast[toUser] = msgID
+	}
```
Comment (42wim) [inline mm-go-irckit/server_commands.go:589]:
```suggestion
		return false
	}
	
	u.msgLastMutex.Lock()
	defer u.msgLastMutex.Unlock()
	
	u.msgLast[toUser] = msgID
```

**Review by 42wim — COMMENTED** [review]
Some small readability nitpicks

**Comment by 42wim** [comment]
Looks a lot of change, but it's basically just early returns where possible.
Thanks for this cool feature :)

**Comment by hloeung** [comment]
> Looks a lot of change, but it's basically just early returns where possible.
> Thanks for this cool feature :)

Ah yes, they all seem good to me. I didn't want to make too many changes blowing out the diff. Thanks for reviewing, let me try convince github to commit those.

**Comment by 42wim** [comment]
sorry, seems my suggestions broke it :(

**Comment by hloeung** [comment]
> sorry, seems my suggestions broke it :(

All good, pushed up a fix. Built and am quickly testing :)

**Comment by hloeung** [comment]
LGTM

    |02:48 -!- Irssi: Join to #haw-test was synced in 21 secs
    |02:48 <hloeung> Test
    |02:48 <hloeung> ... [@@51fo…]
    |02:48 <hloeung> s//Update last
    |02:48 <hloeung> s/51foucr9gffi3mditz61djb5rh/Update by msg ID
    |02:48 <hloeung> Update by msg ID (edited) [@@51fo…]
    |02:49 <hloeung> s//update last, this is @@51fo
    |02:49 <hloeung> update last, this is @@51fo (edited) [@@51fo…]
    |02:49 <hloeung> Test, will be deleted
    |02:49 <hloeung> s//
    |02:49 <hloeung> @@51foucr9gffi3mditz61djb5rh replying to thread
    |02:49 <hloeung> s//Update reply to thread...

13QwO6sAZc

**Comment by hloeung** [comment]
View from my IRC client, irssi:

**Comment by 42wim** [comment]
🎉

**Comment by hloeung** [comment]
> View from my IRC client, irssi:
> 
> MPnoOZ8mVO

Oh FWIW, the shortened message IDs, that's with this irssi script which I just pushed to GitHub - https://github.com/hloeung/matterircd-complete (does message/thread ID tab completion, key binding for most recent message/thread IDs and more...)

**Pull request merged**